smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/rendering/endpoints/EndpointParametersGenerator.kt

```diff
@@ -0,0 +1,205 @@
+        writer.withBlock("public class #L private constructor(builder: Builder) {", "}", CLASS_NAME) {
+            renderFields()
+            renderCompanionObject()
+            write("")
+            if (params.any(KotlinEndpointParameter::isRequired)) {
+                renderInit()
+                write("")
+            }
+            renderEquals()
+            write("")
+            renderHashCode()
+            write("")
+            renderToString()
+            write("")
+            renderCopy()
+            write("")
+            renderBuilder()
+        }
+    }
+
+    private fun renderFields() {
+        params.forEach {
+            writer.ensureSuppressDeprecation(it)
+            it.renderDeclaration(writer, "builder.${it.name}")
+            writer.write("")
+        }
+    }
+
+    private fun renderDocumentation() {
+        writer.dokka {
+            write("The set of values necessary for endpoint resolution.")
+        }
+    }
+
+    private fun renderCompanionObject() {
+        writer.withBlock("public companion object {", "}") {
+            write("public operator fun invoke(block: Builder.() -> Unit): #L = Builder().apply(block).build()", CLASS_NAME)
+        }
+    }
+
+    private fun renderInit() {
+        writer.withBlock("init {", "}") {
+            params.forEach {
+                if (it.isRequired) {
+                    write("""requireNotNull(#1L) { "endpoint provider parameter #1L is required" }""", it.name)
+                }
+            }
+        }
+    }
+
+    private fun renderEquals() {
+        writer.withBlock("public override fun equals(other: Any?): Boolean {", "}") {
+            write("if (this === other) return true")
+            write("if (other !is #L) return false", CLASS_NAME)
+            params.forEach {
+                ensureSuppressDeprecation(it)
+                write("if (this.#1L != other.#1L) return false", it.name)
+            }
+            write("return true")
+        }
+    }
+
+    private fun renderHashCode() {
+        writer.withBlock("public override fun hashCode(): Int {", "}") {
+            if (params.isEmpty()) {
+                write("return 0")
+                return@withBlock
+            }
+
+            ensureSuppressDeprecation(params[0])
+            write("var result = #L?.hashCode() ?: 0", params[0].name)
+            params.drop(1).forEach {
+                ensureSuppressDeprecation(it)
+                write("result = 31 * result + (#L?.hashCode() ?: 0)", it.name)
+            }
+            write("return result")
+        }
+    }
+
+    private fun renderToString() {
+        writer.withBlock("public override fun toString(): String = buildString {", "}") {
+            write("append(\"#L(\")", CLASS_NAME)
+            params.forEachIndexed { index, it ->
+                ensureSuppressDeprecation(it)
+                write(
+                    if (index < params.size - 1) "append(\"#1L=$#1L,\")" else "append(\"#1L=$#1L\")",
```

Comment:
**Style**: Consider pulling the comma out to a dedicated `val` and `write` argument instead of switching directly in the template argument:

```kotlin
val separator = if (index < sortedMembers.size - 1) "," else ")"
...
write("""append("#1L=$#1L#2L")""", it.name, separator)
```